doc/exceptions.md

```diff
@@ -0,0 +1,365 @@
+Always do this.
+```
+
+#### End Statement
+
+The `end` statement ends the handler.
+
+Code following it is reached only if any raised exception is rescued.
+
+#### Begin-Less \Exception Handlers
+
+As seen above, an exception handler may be implemented with `begin` and `end`.
+
+An exception handler may also be implemented as:
+
+- A method body:
+
+    ```
+    def foo(boom: false) # Serves as beginning of exception handler.
+      puts 'Begin.'
+      raise 'Boom!' if boom
+    rescue
+      puts 'Rescued an exception!'
+    else
+      puts 'No exception raised.'
+    end                  # Serves as end of exception handler.
+    ```
+
+- A block:
+
+    ```
+    Dir.chdir('.') do |dir| # Serves as beginning of exception handler.
+      raise 'Boom!'
+    rescue
+      puts 'Rescued an exception!'
+    end                     # Serves as end of exception handler.
+    ```
+
+#### Re-Raising an \Exception
+
+It can be useful to rescue an exception, but allow its eventual effect;
+for example, a program can rescue an exception, log data about it,
+and then "reinstate" the exception.
+
+This may be done via the `raise` method, but in a special way;
+a rescuing clause:
+
+  - Captures an exception.
+  - Does whatever is needed concerning the exception (such as logging it).
+  - Calls method `raise` with no argument,
+    which raises the rescued exception:
+
+```
+begin
+  1 / 0
+rescue ZeroDivisionError
+  # Do needful things (like logging).
+  raise # Raised exception will be ZeroDivisionError, not RuntimeError.
+end
+```
+
+Output:
+
+```
+ruby t.rb
+t.rb:2:in `/': divided by 0 (ZeroDivisionError)
+        from t.rb:2:in `<main>'
+```
+
+#### Retrying
+
+It can be useful to retry a begin clause;
+for example, if it must access a possibly-volatile resource
+(such as a web page),
+it can be useful to try the access more than once
+(in the hope that it may become available):
+
+```
+require 'open-uri'
+begin
+  retries ||= 0
+  tempfile = URI.open("https://docs.ruby-lang.org/en/master/Exception.html")
+  # Do something with the retrieved data.
+rescue OpenURI::HTTPError
+  retry if (retries += 1) < 5
+else
+  puts "Got it, with #{retries} retries."
+end
+```
+
+```
+Got it, with 0 retries.
+```
+
+Note that the retry re-executes the entire begin clause,
+not just the part after the point of failure.
+
+## Raising an \Exception
+
+Raise an exception with method Kernel#raise.
+
+## Custom Exceptions
+
+To provide additional or alternate information,
+you may create custom exception classes;
+each should be a subclass of one of the built-in exception classes.
+
+If you are building a library or gem (or even if you're not),
+it's good practice to start with a single “generic” exception class
+(commonly an immediate subclass of StandardError or RuntimeError),
+and have its other exception classes derive from that class.
+This allows an exception handler to rescue the generic exception,
+thus also rescuing all its derived exceptions.
```

Comment:
I'm not sure if this is really a common or necessarily encouraged practice. Do have an example of where this is done?

Reply:
Net::HTTP has most of its exceptions subclassed from `HTTPClientError` and `HTTPServerError`.

Reply:
I feel like Net::HTTP is a special case, since all HTTP errors are of the 4xx class or 5xx class (for client and server errors, respectively). I don't think this this statement is generally true though.

Reply:
Removed.